File: jarvis/idioma.py
"""Idiomas de Jarvis: español latino y la versión original en inglés.

El idioma se puede cambiar en marcha ("Jarvis, habla en inglés"): la voz, el oído y el
cerebro consultan `actual()` en cada frase.
"""

from dataclasses import dataclass

from . import config


@dataclass(frozen=True)
class Idioma:
    codigo: str
    nombre: str
    voz: str
    tono: str
    velocidad: str
    tratamiento: str
    regla: str                 # cómo debe hablar el cerebro
    saludos: tuple[str, str, str]  # mañana, tarde, noche
    en_linea: str
    despedida: str
    despedidas: frozenset[str]  # frases (normalizadas) que cierran la conversación
    sin_cerebro: str
    confundido: str
    temporizador: str
    temporizador_motivo: str

    def decir(self, plantilla: str, **datos) -> str:
        return plantilla.format(t=self.tratamiento, T=self.tratamiento.capitalize(), **datos)


IDIOMAS = {
    "es": Idioma(
        codigo="es", nombre="español",
        voz=config.VOZ_ES, tono=config.TONO_ES, velocidad=config.VELOCIDAD_ES,
        tratamiento=config.TRATAMIENTO_ES,
        regla="Habla siempre en español latinoamericano neutro y trata al usuario de usted, "
              "con la cortesía de un mayordomo.",
        saludos=("Buenos días", "Buenas tardes", "Buenas noches"),
        en_linea="Todos los sistemas en línea.",
        despedida="Aquí estaré, {t}.",
        despedidas=frozenset({"eso es todo", "nada mas", "nada", "gracias eso es todo",
                              "adios", "hasta luego"}),
        sin_cerebro="Lo siento, {t}, ahora mismo no tengo ningún cerebro disponible. Revise la "
                    "conexión a internet, la clave de Claude o que Ollama esté en marcha.",
        confundido="Me temo que me he confundido con esa petición. ¿Podría pedírmela de otra forma?",
        temporizador="{T}, el temporizador ha terminado.",
        temporizador_motivo="{T}, es la hora: {motivo}.",
    ),
    "en": Idioma(
        codigo="en", nombre="inglés",
        voz=config.VOZ_EN, tono=config.TONO_EN, velocidad=config.VELOCIDAD_EN,
        tratamiento=config.TRATAMIENTO_EN,
        regla="Always reply in English, with the refined British manner of the original JARVIS, "
              "even though these instructions are written in Spanish.",
        saludos=("Good morning", "Good afternoon", "Good evening"),
        en_linea="All systems online.",
        despedida="I'll be here, {t}.",
        despedidas=frozenset({"thats all", "that is all", "nothing", "goodbye", "bye",
                              "thank you thats all"}),
        sin_cerebro="I'm sorry, {t}, I have no brain available at the moment. Please check the "
                    "internet connection, the Claude API key, or that Ollama is running.",
        confundido="I'm afraid I got rather tangled up with that request. Could you phrase it differently?",
        temporizador="{T}, your timer is up.",
        temporizador_motivo="{T}, it's time: {motivo}.",
    ),
}

_actual = IDIOMAS.get(config.IDIOMA, IDIOMAS["es"])


def actual() -> Idioma:
    return _actual


def cambiar(codigo: str) -> Idioma:
    global _actual
    _actual = IDIOMAS[codigo]
    return _actual

File: android/app/src/main/java/com/pabl0sk1/jarvis/Idioma.kt
package com.pabl0sk1.jarvis

/** Español latino e inglés, como en el Jarvis del PC (jarvis/idioma.py). */
data class Idioma(
    val codigo: String,
    val nombre: String,
    val voz: String,
    val tono: String,
    val tratamiento: String,
    val reconocimiento: String,  // idioma para el reconocedor de voz de Android
    val regla: String,
    val saludos: List<String>,   // mañana, tarde, noche
    val enLinea: String,
    val despedida: String,
    val despedidas: Set<String>,
    val sinCerebro: String,
    val confundido: String,
    val temporizador: String,
    val temporizadorMotivo: String,
    val bienvenidaAuto: String,
    val llamadaEntrante: String,
) {
    fun decir(plantilla: String, motivo: String = ""): String = plantilla
        .replace("{t}", tratamiento)
        .replace("{T}", tratamiento.replaceFirstChar { it.uppercase() })
        .replace("{motivo}", motivo)
}

object Idiomas {
    val ES = Idioma(
        codigo = "es", nombre = "español",
        voz = BuildConfig.VOZ_ES, tono = BuildConfig.TONO_ES, tratamiento = "señor",
        reconocimiento = "es-419",
        regla = "Habla siempre en español latinoamericano neutro y trata al usuario de usted, " +
            "con la cortesía de un mayordomo.",
        saludos = listOf("Buenos días", "Buenas tardes", "Buenas noches"),
        enLinea = "Todos los sistemas en línea.",
        despedida = "Aquí estaré, {t}.",
        despedidas = setOf("eso es todo", "nada mas", "nada", "gracias eso es todo", "adios", "hasta luego"),
        sinCerebro = "Lo siento, {t}, ahora mismo no tengo ningún cerebro disponible. " +
            "Revise la conexión a internet.",
        confundido = "Me temo que me he confundido con esa petición. ¿Podría pedírmela de otra forma?",
        temporizador = "{T}, el temporizador ha terminado.",
        temporizadorMotivo = "{T}, es la hora: {motivo}.",
        bienvenidaAuto = "Bienvenido a bordo, {t}. Estoy conectado al auto.",
        llamadaEntrante = "{T}, le llama {motivo}.",
    )

    val EN = Idioma(
        codigo = "en", nombre = "inglés",
        voz = BuildConfig.VOZ_EN, tono = BuildConfig.TONO_EN, tratamiento = "sir",
        reconocimiento = "en-GB",
        regla = "Always reply in English, with the refined British manner of the original JARVIS, " +
            "even though these instructions are written in Spanish.",
        saludos = listOf("Good morning", "Good afternoon", "Good evening"),
        enLinea = "All systems online.",
        despedida = "I'll be here, {t}.",
        despedidas = setOf("thats all", "that is all", "nothing", "goodbye", "bye", "thank you thats all"),
        sinCerebro = "I'm sorry, {t}, I have no brain available at the moment. " +
            "Please check the internet connection.",
        confundido = "I'm afraid I got rather tangled up with that request. Could you phrase it differently?",
        temporizador = "{T}, your timer is up.",
        temporizadorMotivo = "{T}, it's time: {motivo}.",
        bienvenidaAuto = "Welcome aboard, {t}. I'm connected to the car.",
        llamadaEntrante = "{T}, you have a call from {motivo}.",
    )

    val todos = mapOf("es" to ES, "en" to EN)

    @Volatile
    var actual: Idioma = ES
}
